lib/src/revset.rs

```diff
@@ -752,6 +771,22 @@ pub fn expect_string_pattern(node: &ExpressionNode) -> Result<StringPattern, Rev
     revset_parser::expect_pattern_with("string pattern", node, parse_pattern)
 }
 
+fn expect_date_pattern(
+    node: &ExpressionNode,
+    context: &DatePatternContext,
+) -> Result<DatePattern, RevsetParseError> {
+    revset_parser::expect_pattern_with("date pattern", node, |value, kind| match kind {
+        None => Err(RevsetParseError::expression(
+            "Date pattern must specify 'after' or 'before'",
+            node.span,
+        )),
+        Some(kind) => context.to_date_pattern(value, kind).map_err(|err| {
+            RevsetParseError::expression(format!("Unable to parse date pattern: {err}"), node.span)
+                .with_source(err)
```

Comment:
This duplicates error messages. We have to manually convert the error to `Box<dyn _>`.
```rust
        |value, kind| -> Result<_, Box<dyn std::error::Error + Send + Sync>> {
            match kind {
                None => Err("Date pattern must specify 'after' or 'before'".into()),
                Some(kind) => Ok(context.to_date_pattern(value, kind)?),
            }
        }
```